File: backend/solver.py
from __future__ import annotations

from random import Random
from typing import Dict, List, Tuple, Any

BOARD_SIZE = 6
ITEM_TYPES = [
    "beer",
    "liquor",
    "keg",
    "food",
    "snacks",
    "decor",
    "speakers",
    "drugs",
    "girls",
    "guys",
    "cups",
    "trashcans",
    "music",
]

ACTIVATION_DEPENDENCIES = {
    "beer": "cups",
    "liquor": "cups",
    "keg": "cups",
    "food": "trashcans",
    "snacks": "trashcans",
    "speakers": "music",
}
ACTIVATION_PERKS = {
    "beerNoCups": "beer",
    "liquorNoCups": "liquor",
    "foodNoTrashcan": "food",
}


def neighbors(x: int, y: int) -> List[Tuple[int, int]]:
    out = []
    for dy in (-1, 0, 1):
        for dx in (-1, 0, 1):
            if dx == 0 and dy == 0:
                continue
            nx, ny = x + dx, y + dy
            if 0 <= nx < BOARD_SIZE and 0 <= ny < BOARD_SIZE:
                out.append((nx, ny))
    return out


def dependency_for(item_type: str, activation_perks: Dict[str, bool]) -> str | None:
    dep = ACTIVATION_DEPENDENCIES.get(item_type)
    if not dep:
        return None
    for perk_name, perk_type in ACTIVATION_PERKS.items():
        if perk_type == item_type and activation_perks.get(perk_name):
            return None
    return dep


def evaluate(payload: Dict[str, Any]) -> Dict[str, Any]:
    board = payload["board"]
    prices = payload.get("prices", {})
    perks = payload.get("perks", {})
    activation_perks = perks.get("activation", {})

    items = {(it["x"], it["y"]): it for it in board.get("items", [])}
    active = {}
    global_counts = {t: 0 for t in ITEM_TYPES}

    for it in board.get("items", []):
        dep = dependency_for(it["type"], activation_perks)
        is_active = True
        if dep:
            is_active = any(items.get((nx, ny), {}).get("type") == dep for nx, ny in neighbors(it["x"], it["y"]))
        active[(it["x"], it["y"])] = is_active
        if is_active:
            global_counts[it["type"]] += 1

    satisfied = 0
    diagnostics = []
    for ch in board.get("characters", []):
        local_diag = []
        for need in ch.get("localNeeds", []):
            count = 0
            for nx, ny in neighbors(ch["x"], ch["y"]):
                it = items.get((nx, ny))
                if it and it["type"] == need["type"] and active.get((nx, ny), False):
                    count += 1
            ok = count >= int(need["required"])
            satisfied += int(ok)
            local_diag.append({**need, "count": count, "satisfied": ok})

        g = ch["globalNeed"]
        g_count = global_counts.get(g["type"], 0)
        ok = g_count == int(g["required"]) if g["operator"] == "eq" else g_count >= int(g["required"])
        satisfied += int(ok)
        diagnostics.append({"characterId": ch["id"], "local": local_diag, "global": {**g, "count": g_count, "satisfied": ok}})

    spent = sum(prices.get(it["type"], 0) for it in board.get("items", []))
    return {
        "satisfiedNeeds": satisfied,
        "multiplier": 1.0 + 0.25 * satisfied,
        "spent": spent,
        "globalCounts": global_counts,
        "activeItems": sum(1 for it in board.get("items", []) if active.get((it["x"], it["y"]), False)),
        "activeByCell": {f"{x},{y}": v for (x, y), v in active.items()},
        "diagnostics": diagnostics,
    }


def optimize(payload: Dict[str, Any]) -> Dict[str, Any]:
    mode = payload.get("mode", "unlimited")
    budget = payload.get("budget", 0)
    prices = payload.get("prices", {})
    iterations = int(payload.get("iterations", 5000))
    rnd = Random(int(payload.get("seed", 1234)))

    current = {
        "board": payload["board"],
        "perks": payload.get("perks", {}),
        "prices": prices,
    }
    current_eval = evaluate(current)
    best = {"board": {"characters": [*current["board"]["characters"]], "items": [*current["board"]["items"]]}, "perks": current["perks"], "prices": prices}
    best_eval = current_eval

    candidate_types = {"cups", "trashcans", "music"}
    for ch in current["board"]["characters"]:
        candidate_types.update(n["type"] for n in ch["localNeeds"])
        candidate_types.add(ch["globalNeed"]["type"])
    candidate_types = [t for t in candidate_types if t]

    def clone_state(st: Dict[str, Any]) -> Dict[str, Any]:
        return {
            "board": {
                "characters": [dict(c) | {"localNeeds": [dict(n) for n in c["localNeeds"]], "globalNeed": dict(c["globalNeed"])} for c in st["board"]["characters"]],
                "items": [dict(it) for it in st["board"]["items"]],
            },
            "perks": st["perks"],
            "prices": st["prices"],
        }

    def score(ev: Dict[str, Any], item_count: int) -> Tuple[int, int, int]:
        return (ev["satisfiedNeeds"], -ev["spent"], -item_count)

    for _ in range(iterations):
        nxt = clone_state(current)
        items = nxt["board"]["items"]
        op = rnd.random()
        if op < 0.34 and items:
            items.pop(rnd.randrange(len(items)))
        elif op < 0.68:
            occupied = {(c["x"], c["y"]) for c in nxt["board"]["characters"]} | {(i["x"], i["y"]) for i in items}
            free = [(x, y) for y in range(BOARD_SIZE) for x in range(BOARD_SIZE) if (x, y) not in occupied]
            if free:
                x, y = free[rnd.randrange(len(free))]
                items.append({"id": f"it-{rnd.random()}", "x": x, "y": y, "type": candidate_types[rnd.randrange(len(candidate_types))]})
        elif items:
            idx = rnd.randrange(len(items))
            if rnd.random() < 0.5:
                items[idx]["type"] = candidate_types[rnd.randrange(len(candidate_types))]
            else:
                occupied = {(c["x"], c["y"]) for c in nxt["board"]["characters"]} | {(i["x"], i["y"]) for j, i in enumerate(items) if j != idx}
                free = [(x, y) for y in range(BOARD_SIZE) for x in range(BOARD_SIZE) if (x, y) not in occupied]
                if free:
                    items[idx]["x"], items[idx]["y"] = free[rnd.randrange(len(free))]

        nxt_eval = evaluate(nxt)
        if mode == "budgeted" and nxt_eval["spent"] > budget:
            continue

        if score(nxt_eval, len(items)) > score(current_eval, len(current["board"]["items"])) or rnd.random() < 0.08:
            current, current_eval = nxt, nxt_eval
        if score(nxt_eval, len(items)) > score(best_eval, len(best["board"]["items"])):
            best, best_eval = clone_state(nxt), nxt_eval

    return {"board": best["board"], "metrics": best_eval}

File: solver.js
(function () {
  const BOARD_SIZE = 6;
  const ITEM_TYPES = [
    'beer','liquor','keg','food','snacks','decor','speakers','drugs','girls','guys','cups','trashcans','music'
  ];

  const ACTIVATION_DEPENDENCIES = {
    beer: 'cups',
    liquor: 'cups',
    keg: 'cups',
    food: 'trashcans',
    snacks: 'trashcans',
    speakers: 'music',
  };

  const ACTIVATION_PERK_MAP = {
    beerNoCups: 'beer',
    liquorNoCups: 'liquor',
    foodNoTrashcan: 'food',
  };

  const LOCAL_TYPES = ITEM_TYPES.filter((t) => !['cups','trashcans','music'].includes(t));

  function key(x, y) { return `${x},${y}`; }

  function inBounds(x, y) { return x >= 0 && y >= 0 && x < BOARD_SIZE && y < BOARD_SIZE; }

  function neighbors(x, y) {
    const out = [];
    for (let dy = -1; dy <= 1; dy++) {
      for (let dx = -1; dx <= 1; dx++) {
        if (!dx && !dy) continue;
        const nx = x + dx;
        const ny = y + dy;
        if (inBounds(nx, ny)) out.push({ x: nx, y: ny });
      }
    }
    return out;
  }

  function dependencyFor(type, perks) {
    const dep = ACTIVATION_DEPENDENCIES[type];
    if (!dep) return null;
    for (const perkName of Object.keys(ACTIVATION_PERK_MAP)) {
      if (ACTIVATION_PERK_MAP[perkName] === type && perks[perkName]) {
        return null;
      }
    }
    return dep;
  }

  function buildMaps(board) {
    const chars = new Map();
    const items = new Map();
    for (const c of board.characters) chars.set(key(c.x, c.y), c);
    for (const i of board.items) items.set(key(i.x, i.y), i);
    return { chars, items };
  }

  function isItemActive(item, itemsMap, perks) {
    const dep = dependencyFor(item.type, perks);
    if (!dep) return true;
    return neighbors(item.x, item.y).some(({ x, y }) => {
      const other = itemsMap.get(key(x, y));
      return other && other.type === dep;
    });
  }

  function evaluateBoard(state) {
    const { board, perks } = state;
    const { items } = buildMaps(board);
    const activeByCell = new Map();
    const globalCounts = Object.fromEntries(ITEM_TYPES.map((t) => [t, 0]));

    for (const it of board.items) {
      const active = isItemActive(it, items, perks.activation);
      activeByCell.set(key(it.x, it.y), active);
      if (active) globalCounts[it.type] += 1;
    }

    let satisfied = 0;
    const diagnostics = [];

    for (const ch of board.characters) {
      const localDetails = ch.localNeeds.map((need) => {
        let count = 0;
        for (const n of neighbors(ch.x, ch.y)) {
          const it = items.get(key(n.x, n.y));
          if (!it) continue;
          if (it.type !== need.type) continue;
          if (!activeByCell.get(key(n.x, n.y))) continue;
          count += 1;
        }
        const ok = count >= need.required;
        if (ok) satisfied += 1;
        return { ...need, count, satisfied: ok };
      });

      const g = ch.globalNeed;
      const count = globalCounts[g.type] || 0;
      const ok = g.operator === 'eq' ? count === g.required : count >= g.required;
      if (ok) satisfied += 1;

      diagnostics.push({
        characterId: ch.id,
        local: localDetails,
        global: { ...g, count, satisfied: ok },
      });
    }

    const multiplier = 1 + 0.25 * satisfied;
    const spent = board.items.reduce((acc, it) => acc + (state.prices[it.type] ?? 0), 0);

    return {
      satisfiedNeeds: satisfied,
      multiplier,
      spent,
      activeByCell,
      diagnostics,
      globalCounts,
      activeItems: board.items.filter((it) => activeByCell.get(key(it.x, it.y))).length,
    };
  }

  function freeCells(board) {
    const blocked = new Set();
    for (const c of board.characters) blocked.add(key(c.x, c.y));
    for (const i of board.items) blocked.add(key(i.x, i.y));
    const out = [];
    for (let y = 0; y < BOARD_SIZE; y++) {
      for (let x = 0; x < BOARD_SIZE; x++) {
        if (!blocked.has(key(x, y))) out.push({ x, y });
      }
    }
    return out;
  }

  function mulberry32(seed) {
    let t = seed >>> 0;
    return function () {
      t += 0x6D2B79F5;
      let r = Math.imul(t ^ t >>> 15, t | 1);
      r ^= r + Math.imul(r ^ r >>> 7, r | 61);
      return ((r ^ r >>> 14) >>> 0) / 4294967296;
    };
  }

  function optimize(initial, options = {}) {
    const iterations = options.iterations ?? 2500;
    const rng = mulberry32(options.seed ?? 1234);
    const budget = initial.mode === 'budgeted' ? initial.budget : Infinity;

    function score(evalRes) {
      return [
        evalRes.satisfiedNeeds,
        -evalRes.spent,
        -initial.board.items.length,
      ];
    }

    function better(a, b) {
      for (let i = 0; i < a.length; i++) {
        if (a[i] > b[i]) return true;
        if (a[i] < b[i]) return false;
      }
      return false;
    }

    let current = structuredClone(initial);
    let currentEval = evaluateBoard(current);
    let best = structuredClone(current);
    let bestEval = currentEval;

    const candidatePool = new Set(['cups', 'trashcans', 'music']);
    for (const ch of initial.board.characters) {
      ch.localNeeds.forEach((n) => candidatePool.add(n.type));
      candidatePool.add(ch.globalNeed.type);
    }
    const candidates = [...candidatePool].filter(Boolean);

    for (let step = 0; step < iterations; step++) {
      const next = structuredClone(current);
      const op = rng();

      if (op < 0.34 && next.board.items.length) {
        next.board.items.splice(Math.floor(rng() * next.board.items.length), 1);
      } else if (op < 0.68) {
        const free = freeCells(next.board);
        if (free.length) {
          const pos = free[Math.floor(rng() * free.length)];
          const type = candidates[Math.floor(rng() * candidates.length)];
          next.board.items.push({ id: crypto.randomUUID(), ...pos, type });
        }
      } else if (next.board.items.length) {
        const idx = Math.floor(rng() * next.board.items.length);
        if (rng() < 0.5) {
          const type = candidates[Math.floor(rng() * candidates.length)];
          next.board.items[idx].type = type;
        } else {
          const occupied = new Set(next.board.characters.map((c) => key(c.x, c.y)));
          next.board.items.forEach((it, i) => { if (i !== idx) occupied.add(key(it.x, it.y)); });
          const spots = [];
          for (let y = 0; y < BOARD_SIZE; y++) for (let x = 0; x < BOARD_SIZE; x++) if (!occupied.has(key(x, y))) spots.push({ x, y });
          if (spots.length) {
            const spot = spots[Math.floor(rng() * spots.length)];
            next.board.items[idx].x = spot.x;
            next.board.items[idx].y = spot.y;
          }
        }
      }

      const nextEval = evaluateBoard(next);
      if (nextEval.spent > budget) continue;

      const nScore = score(nextEval);
      const cScore = score(currentEval);
      if (better(nScore, cScore) || rng() < 0.08) {
        current = next;
        currentEval = nextEval;
      }
      const bScore = score(bestEval);
      if (better(nScore, bScore)) {
        best = structuredClone(next);
        bestEval = nextEval;
      }
    }

    return {
      board: best.board,
      metrics: bestEval,
    };
  }

  window.PartyPlannerSolver = {
    BOARD_SIZE,
    ITEM_TYPES,
    LOCAL_TYPES,
    ACTIVATION_DEPENDENCIES,
    evaluateBoard,
    optimize,
  };
})();
